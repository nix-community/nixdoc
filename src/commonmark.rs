// Copyright (C) 2018 Vincent Ambo <mail@tazj.in>
//
// nixdoc is free software: you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

//! This module implements CommonMark output for a struct
//! representing a single entry in the manual.

use serde::Serialize;

/// Represent a single function argument name and its (optional)
/// doc-string.
#[derive(Clone, Debug, Serialize)]
pub struct SingleArg {
    pub name: String,
    pub doc: Option<String>,
}

/// Represent a function argument, which is either a flat identifier
/// or a pattern set.
#[derive(Clone, Debug, Serialize)]
pub enum Argument {
    /// Flat function argument (e.g. `n: n * 2`).
    Flat(SingleArg),

    /// Pattern function argument (e.g. `{ name, age }: ...`)
    Pattern(Vec<SingleArg>),
}

impl Argument {
    /// Write CommonMark structure for a single function argument.
    /// We use the definition list extension, which prepends each argument with `: `.
    /// For pattern arguments, we create a nested definition list.
    fn format_argument(self) -> String {
        match self {
            // Write a flat argument entry, e.g. `id = x: x`
            //
            // `x`
            // : Function argument
            Argument::Flat(arg) => {
                format!(
                    "`{}`\n\n: {}\n\n",
                    arg.name,
                    handle_indentation(arg.doc.unwrap_or("Function argument".into()).trim())
                )
            }

            // Write a pattern argument entry and its individual
            // parameters as a nested structure, e.g.:
            //
            // `foo = { a }: a`
            //
            // structured function argument
            // : `a`
            //   : Function argument
            Argument::Pattern(pattern_args) => {
                let mut inner = String::new();
                for pattern_arg in pattern_args {
                    inner += &Argument::Flat(pattern_arg).format_argument();
                }

                let indented = textwrap::indent(&inner, "  ");

                format!(
                    // The `:` creates another definition list of which `indented` is the term.
                    "structured function argument\n\n: {}",
                    // drop leading indentation on the first line, the `: ` serves this function
                    // already.
                    indented.trim_start()
                )
            }
        }
    }
}

/// Since the first line starts with `: `, indent every other line by 2 spaces, so
/// that the text aligns, to result in:
///
/// : first line
///   every other line
fn handle_indentation(raw: &str) -> String {
    match raw.split_once('\n') {
        Some((first, rest)) => {
            format!("{}\n{}", first, textwrap::indent(rest, "  "))
        }
        None => raw.into(),
    }
}

/// Generate the identifier for CommonMark.
/// ident is used as URL Encoded link to the function and has thus stricter rules (i.e. "' " in "lib.map' "  is not allowed).
pub(crate) fn get_identifier(prefix: &String, category: &String, name: &String) -> String {
    let name_prime = name.replace('\'', "-prime");
    vec![prefix, category, &name_prime]
        .into_iter()
        .filter(|x| !x.is_empty())
        .cloned()
        .collect::<Vec<String>>()
        .join(".")
}

/// Generate the title for CommonMark.
/// the title is the human-readable name of the function.
pub(crate) fn get_title(prefix: &String, category: &String, name: &String) -> String {
    vec![prefix, category, name]
        .into_iter()
        .filter(|x| !x.is_empty())
        .cloned()
        .collect::<Vec<String>>()
        .join(".")
}

/// Represents a single manual section describing a library function.
#[derive(Clone, Debug, Serialize)]
pub struct ManualEntry {
    /// Prefix for the category (e.g. 'lib' or 'utils').
    pub prefix: String,

    /// Name of the function category (e.g. 'strings', 'trivial', 'attrsets').
    pub category: String,

    /// Location of the function.
    pub location: Option<String>,

    /// Name of the section (used as the title).
    pub name: String,

    /// Type signature (if provided). This is not actually a checked
    /// type signature in any way.
    pub fn_type: Option<String>,

    /// Primary description of the entry. Each entry is written as a
    /// separate paragraph.
    pub description: Vec<String>,

    /// Usage example for the entry.
    pub example: Option<String>,

    /// Arguments of the function.
    pub args: Vec<Argument>,
}

impl ManualEntry {
    pub(crate) fn get_ident_title(&self) -> (String, String) {
        let ident = get_identifier(&self.prefix, &self.category, &self.name);
        let title = get_title(&self.prefix, &self.category, &self.name);
        (ident, title)
    }

    /// Write a single CommonMark entry for a documented Nix function.
    ///
    /// # Arguments
    ///
    /// - `anchor_prefix`: The prefix to use for the anchor links.
    ///   In Nixpkgs this would be "function-library-".
    /// - `output`: The output string to append the CommonMark onto.
    pub fn write_section(self, anchor_prefix: &str, output: &mut String) -> String {
        let (ident, title) = self.get_ident_title();
        output.push_str(&format!(
            "## `{}` {{#{}{}}}\n\n",
            title, anchor_prefix, ident
        ));

        // <subtitle> (type signature)
        if let Some(t) = &self.fn_type {
            if t.lines().count() > 1 {
                output.push_str(&format!("**Type**:\n```\n{}\n```\n\n", t));
            } else {
                output.push_str(&format!("**Type**: `{}`\n\n", t));
            }
        }

        // Primary doc string
        // TODO: Split paragraphs?
        for paragraph in &self.description {
            output.push_str(&format!("{}\n\n", paragraph));
        }

        // Function argument names
        if !self.args.is_empty() {
            for arg in self.args {
                output.push_str(&format!("{}\n", arg.format_argument()));
            }
        }

        // Example program listing (if applicable)
        //
        // TODO: In grhmc's version there are multiple (named)
        // examples, how can this be achieved automatically?
        if let Some(example) = &self.example {
            output.push_str(&format!(
                "::: {{.example #{}example-{}}}\n",
                anchor_prefix, ident
            ));
            output.push_str(&format!("# `{}` usage example\n\n", title));
            output.push_str(&format!("```nix\n{}\n```\n:::\n\n", example.trim()));
        }

        if let Some(loc) = self.location {
            output.push_str(&String::from(format!("Located at {loc}.\n\n")));
        }

        output.to_string()
    }
}
