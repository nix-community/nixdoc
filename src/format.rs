use textwrap::dedent;

/// Ensure all lines in a multi-line doc-comments have the same indentation.
///
/// Consider such a doc comment:
///
/// ```nix
/// {
///   /* foo is
///   the value:
///     10
///   */
///   foo = 10;
/// }
/// ```
///
/// The parser turns this into:
///
/// ```
/// foo is
///   the value:
///     10
/// ```
///
///
/// where the first line has no leading indentation, and all other lines have preserved their
/// original indentation.
///
/// What we want instead is:
///
/// ```
/// foo is
/// the value:
///   10
/// ```
///
/// i.e. we want the whole thing to be dedented. To achieve this, we remove all leading whitespace
/// from the first line, and remove all common whitespace from the rest of the string.
pub fn handle_indentation(raw: &str) -> Option<String> {
    let result: String = match raw.split_once('\n') {
        Some((first, rest)) => {
            format!("{}\n{}", first.trim_start(), dedent(rest))
        }
        None => raw.into(),
    };

    Some(result.trim().to_owned()).filter(|s| !s.is_empty())
}

/// Shift down markdown headings
///
/// Performs a line-wise matching to '# Heading '
///
/// Counts the current numbers of '#' and adds levels: [usize] to them
/// levels := 1; gives
/// '# Heading' -> '## Heading'
///
/// Commonmark markdown has 6 levels of headings. Everything beyond that (e.g., H7) is not supported and may produce unexpected renderings.
/// by default this function makes sure, headings don't exceed the H6 boundary.
/// levels := 2;
/// ...
/// H4 -> H6
/// H6 -> H6
///
pub fn shift_headings(raw: &str, levels: usize) -> String {
    let mut result = String::new();
    for line in raw.split_inclusive('\n') {
        if line.trim_start().starts_with('#') {
            result.push_str(&handle_heading(line, levels));
        } else {
            result.push_str(line);
        }
    }
    result
}

// Dumb heading parser.
pub fn handle_heading(line: &str, levels: usize) -> String {
    let chars = line.chars();

    // let mut leading_trivials: String = String::new();
    let mut hashes = String::new();
    let mut rest = String::new();
    for char in chars {
        match char {
            '#' if rest.is_empty() => {
                // only collect hashes if no other tokens
                hashes.push(char)
            }
            _ => rest.push(char),
        }
    }
    let new_hashes = match hashes.len() + levels {
        // We reached the maximum heading size.
        6.. => "#".repeat(6),
        _ => "#".repeat(hashes.len() + levels),
    };

    format!("{new_hashes}{rest}")
}
