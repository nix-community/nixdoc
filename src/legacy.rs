use rnix::{
    ast::{AstToken, Comment, Expr, Lambda, Param},
    SyntaxKind, SyntaxNode,
};
use rowan::ast::AstNode;

use crate::{
    commonmark::{Argument, SingleArg},
    format::handle_indentation,
    retrieve_doc_comment,
};

/// Retrieve documentation comments.
pub fn retrieve_legacy_comment(node: &SyntaxNode, allow_line_comments: bool) -> Option<String> {
    // if the current node has a doc comment it'll be immediately preceded by that comment,
    // or there will be a whitespace token and *then* the comment tokens before it. We merge
    // multiple line comments into one large comment if they are on adjacent lines for
    // documentation simplicity.
    let mut token = node.first_token()?.prev_token()?;
    if token.kind() == SyntaxKind::TOKEN_WHITESPACE {
        token = token.prev_token()?;
    }
    if token.kind() != SyntaxKind::TOKEN_COMMENT {
        return None;
    }
    // if we want to ignore line comments (eg because they may contain deprecation
    // comments on attributes) we'll backtrack to the first preceding multiline comment.
    while !allow_line_comments && token.text().starts_with('#') {
        token = token.prev_token()?;
        if token.kind() == SyntaxKind::TOKEN_WHITESPACE {
            token = token.prev_token()?;
        }
        if token.kind() != SyntaxKind::TOKEN_COMMENT {
            return None;
        }
    }

    if token.text().starts_with("/*") {
        return Some(Comment::cast(token)?.text().to_string());
    }

    // backtrack to the start of the doc comment, allowing only adjacent line comments.
    // we don't care much about optimization here, doc comments aren't long enough for that.
    if token.text().starts_with('#') {
        let mut result = String::new();
        while let Some(comment) = Comment::cast(token) {
            if !comment.syntax().text().starts_with('#') {
                break;
            }
            result.insert_str(0, comment.text().trim());
            let ws = match comment.syntax().prev_token() {
                Some(t) if t.kind() == SyntaxKind::TOKEN_WHITESPACE => t,
                _ => break,
            };
            // only adjacent lines continue a doc comment, empty lines do not.
            match ws.text().strip_prefix('\n') {
                Some(trail) if !trail.contains('\n') => result.insert(0, ' '),
                _ => break,
            }
            token = match ws.prev_token() {
                Some(c) => c,
                _ => break,
            };
        }
        return Some(result);
    }

    None
}

/// Traverse directly chained nix lambdas and collect the identifiers of all lambda arguments
/// until an unexpected AST node is encountered.
pub fn collect_lambda_args(mut lambda: Lambda) -> Vec<Argument> {
    let mut args = vec![];

    loop {
        match lambda.param().unwrap() {
            // a variable, e.g. `x:` in `id = x: x`
            // Single args are not supported by RFC145, due to ambiguous placement rules.
            Param::IdentParam(id) => {
                args.push(Argument::Flat(SingleArg {
                    name: id.to_string(),
                    doc: handle_indentation(
                        &retrieve_legacy_comment(id.syntax(), true).unwrap_or_default(),
                    ),
                }));
            }
            // an ident in a pattern, e.g. `a` in `foo = { a }: a`
            Param::Pattern(pat) => {
                // collect doc-comments for each lambda formal
                // Lambda formals are supported by RFC145
                let pattern_vec: Vec<_> = pat
                    .pat_entries()
                    .map(|entry| SingleArg {
                        name: entry.ident().unwrap().to_string(),
                        doc: handle_indentation(
                            &retrieve_doc_comment(entry.syntax(), Some(1))
                                .or(retrieve_legacy_comment(entry.syntax(), true))
                                .unwrap_or_default(),
                        ),
                    })
                    .collect();

                args.push(Argument::Pattern(pattern_vec));
            }
        }

        // Curried or not?
        match lambda.body() {
            Some(Expr::Lambda(inner)) => lambda = inner,
            _ => break,
        }
    }

    args
}
