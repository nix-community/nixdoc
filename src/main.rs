// Copyright (C) 2018 Vincent Ambo <mail@tazj.in>
//
// nixdoc is free software: you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

//! This tool generates CommonMark from a Nix file defining library
//! functions, such as the files in `lib/` in the nixpkgs repository.
//!
//! TODO:
//! * extract function argument names
//! * extract line number & add it to generated output
//! * figure out how to specify examples (& leading whitespace?!)

mod comment;
mod commonmark;
mod format;
mod legacy;
#[cfg(test)]
mod test;

use crate::{format::handle_indentation, legacy::retrieve_legacy_comment};

use self::comment::get_expr_docs;
use self::commonmark::*;
use format::shift_headings;
use legacy::collect_lambda_args;
use rnix::{
    ast::{Attr, AttrpathValue, Expr, Inherit, LetIn},
    SyntaxKind, SyntaxNode,
};
use rowan::{ast::AstNode, WalkEvent};
use std::fs;

use std::collections::HashMap;
use std::io;
use std::io::Write;

use clap::Parser;
use std::path::PathBuf;

/// Command line arguments for nixdoc
#[derive(Debug, Parser)]
#[command(author, version, about)]
struct Options {
    /// Prefix for the category (e.g. 'lib' or 'utils').
    #[arg(short, long, default_value_t = String::from("lib"))]
    prefix: String,

    /// Name of the function category (e.g. 'strings', 'attrsets').
    #[arg(short, long)]
    category: String,

    /// Description of the function category.
    #[arg(short, long)]
    description: String,

    /// Nix file to process.
    #[arg(short, long)]
    file: PathBuf,

    /// Path to a file containing location data as JSON.
    #[arg(short, long)]
    locs: Option<PathBuf>,
}

#[derive(Debug)]
struct DocComment {
    /// Primary documentation string.
    doc: String,

    /// Optional type annotation for the thing being documented.
    /// This is only available as legacy feature
    doc_type: Option<String>,

    /// Usage example(s) (interpreted as a single code block)
    /// This is only available as legacy feature
    example: Option<String>,
}

#[derive(Debug)]
struct DocItem {
    name: String,
    comment: DocComment,
    args: Vec<Argument>,
}

impl DocItem {
    fn into_entry(self, prefix: &str, category: &str) -> ManualEntry {
        ManualEntry {
            prefix: prefix.to_string(),
            category: category.to_string(),
            name: self.name,
            description: self
                .comment
                .doc
                .split("\n\n")
                .map(|s| s.to_string())
                .collect(),
            fn_type: self.comment.doc_type,
            example: self.comment.example,
            args: self.args,
        }
    }
}

/// Returns a rfc145 doc-comment if one is present
pub fn retrieve_doc_comment(node: &SyntaxNode, shift_headings_by: Option<usize>) -> Option<String> {
    let doc_comment = get_expr_docs(node);

    doc_comment.map(|doc_comment| {
        shift_headings(
            &handle_indentation(&doc_comment).unwrap_or(String::new()),
            // H1 to H4 can be used in the doc-comment with the current rendering.
            // They will be shifted to H3, H6
            // H1 and H2 are currently used by the outer rendering. (category and function name)
            shift_headings_by.unwrap_or(2),
        )
    })
}

/// Transforms an AST node into a `DocItem` if it has a leading
/// documentation comment.
fn retrieve_doc_item(node: &AttrpathValue) -> Option<DocItem> {
    let ident = node.attrpath().unwrap();
    // TODO this should join attrs() with '.' to handle whitespace, dynamic attrs and string
    // attrs. none of these happen in nixpkgs lib, and the latter two should probably be
    // rejected entirely.
    let item_name = ident.to_string();

    let doc_comment = retrieve_doc_comment(node.syntax(), Some(2));
    match doc_comment {
        Some(comment) => Some(DocItem {
            name: item_name,
            comment: DocComment {
                doc: comment,
                doc_type: None,
                example: None,
            },
            args: vec![],
        }),
        // Fallback to legacy comment is there is no doc_comment
        None => {
            let comment = retrieve_legacy_comment(node.syntax(), false)?;
            Some(DocItem {
                name: item_name,
                comment: parse_doc_comment(&comment),
                args: vec![],
            })
        }
    }
}

/// Dumb, mutable, hacky doc comment "parser".
fn parse_doc_comment(raw: &str) -> DocComment {
    enum ParseState {
        Doc,
        Type,
        Example,
    }

    let mut state = ParseState::Doc;

    // Split the string into three parts, docs, type and example
    let mut doc_str = String::new();
    let mut type_str = String::new();
    let mut example_str = String::new();

    for line in raw.split_inclusive('\n') {
        let trimmed_line = line.trim();
        if let Some(suffix) = trimmed_line.strip_prefix("Type:") {
            state = ParseState::Type;
            type_str.push_str(suffix);
            type_str.push('\n');
        } else if let Some(suffix) = trimmed_line.strip_prefix("Example:") {
            state = ParseState::Example;
            example_str.push_str(suffix);
            example_str.push('\n');
        } else {
            match state {
                ParseState::Doc => doc_str.push_str(line),
                ParseState::Type => type_str.push_str(line),
                ParseState::Example => example_str.push_str(line),
            }
        }
    }

    DocComment {
        doc: handle_indentation(&doc_str).unwrap_or(String::new()),
        doc_type: handle_indentation(&type_str),
        example: handle_indentation(&example_str),
    }
}

/// Traverse the arena from a top-level SetEntry and collect, where
/// possible:
///
/// 1. The identifier of the set entry itself.
/// 2. The attached doc comment on the entry.
/// 3. The argument names of any curried functions (pattern functions
///    not yet supported).
fn collect_entry_information(entry: AttrpathValue) -> Option<DocItem> {
    let doc_item = retrieve_doc_item(&entry)?;

    if let Some(Expr::Lambda(l)) = entry.value() {
        Some(DocItem {
            args: collect_lambda_args(l),
            ..doc_item
        })
    } else {
        Some(doc_item)
    }
}

// a binding is an assignment, which can take place in an attrset
// - as attributes
// - as inherits
fn collect_bindings(
    node: &SyntaxNode,
    prefix: &str,
    category: &str,
    scope: HashMap<String, ManualEntry>,
) -> Vec<ManualEntry> {
    for ev in node.preorder() {
        match ev {
            WalkEvent::Enter(n) if n.kind() == SyntaxKind::NODE_ATTR_SET => {
                let mut entries = vec![];
                for child in n.children() {
                    if let Some(apv) = AttrpathValue::cast(child.clone()) {
                        entries.extend(
                            collect_entry_information(apv)
                                .map(|di| di.into_entry(prefix, category)),
                        );
                    } else if let Some(inh) = Inherit::cast(child) {
                        // `inherit (x) ...` needs much more handling than we can
                        // reasonably do here
                        if inh.from().is_some() {
                            continue;
                        }
                        entries.extend(inh.attrs().filter_map(|a| match a {
                            Attr::Ident(i) => scope.get(&i.syntax().text().to_string()).cloned(),
                            // ignore non-ident keys. these aren't useful as lib
                            // functions in general anyway.
                            _ => None,
                        }));
                    }
                }
                return entries;
            }
            _ => (),
        }
    }

    vec![]
}

// Main entrypoint for collection
// TODO: document
fn collect_entries(root: rnix::Root, prefix: &str, category: &str) -> Vec<ManualEntry> {
    // we will look into the top-level let and its body for function docs.
    // we only need a single level of scope for this.
    // since only the body can export a function we don't need to implement
    // mutually recursive resolution.
    for ev in root.syntax().preorder() {
        match ev {
            WalkEvent::Enter(n) if n.kind() == SyntaxKind::NODE_LET_IN => {
                return collect_bindings(
                    LetIn::cast(n.clone()).unwrap().body().unwrap().syntax(),
                    prefix,
                    category,
                    n.children()
                        .filter_map(AttrpathValue::cast)
                        .filter_map(collect_entry_information)
                        .map(|di| (di.name.to_string(), di.into_entry(prefix, category)))
                        .collect(),
                );
            }
            WalkEvent::Enter(n) if n.kind() == SyntaxKind::NODE_ATTR_SET => {
                return collect_bindings(&n, prefix, category, Default::default());
            }
            _ => (),
        }
    }

    vec![]
}

fn retrieve_description(nix: &rnix::Root, description: &str, category: &str) -> String {
    format!(
        "# {} {{#sec-functions-library-{}}}\n{}\n",
        description,
        category,
        &nix.syntax()
            .first_child()
            .and_then(|node| retrieve_doc_comment(&node, Some(1))
                .or(retrieve_legacy_comment(&node, false)))
            .and_then(|doc_item| handle_indentation(&doc_item))
            .unwrap_or_default()
    )
}

fn main() {
    let mut output = io::stdout();
    let opts = Options::parse();
    let src = fs::read_to_string(&opts.file).unwrap();
    let locs = match opts.locs {
        None => Default::default(),
        Some(p) => fs::read_to_string(p)
            .map_err(|e| e.to_string())
            .and_then(|json| serde_json::from_str(&json).map_err(|e| e.to_string()))
            .expect("could not read location information"),
    };
    let nix = rnix::Root::parse(&src).ok().expect("failed to parse input");
    let description = retrieve_description(&nix, &opts.description, &opts.category);

    // TODO: move this to commonmark.rs
    writeln!(output, "{}", description).expect("Failed to write header");

    for entry in collect_entries(nix, &opts.prefix, &opts.category) {
        entry
            .write_section(&locs, &mut output)
            .expect("Failed to write section")
    }
}
