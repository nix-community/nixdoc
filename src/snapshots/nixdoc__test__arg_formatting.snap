---
source: src/main.rs
expression: output
---
## `lib.options.fn` {#function-library-lib.options.fn}

documented function

`a`

: single arugment


structured function argument

: `default`

  : documented argument

  `example`

  : i like this argument. another!



